File: Lab5/Python/TPoint.py
from math import *


class TCoordinates:
    __x = 0
    __y = 0
    __z = 0

    def __init__(self, x, y, z):
        self.__x = x
        self.__y = y
        self.__z = z

    @property
    def x(self):
        return self.__x

    @property
    def y(self):
        return self.__y

    @property
    def z(self):
        return self.__z

    @x.setter
    def x(self, value):
        self.__x = value

    @y.setter
    def y(self, value):
        self.__y = value

    @z.setter
    def z(self, value):
        self.__z = value


class TPoint:
    def __init__(self, x, y, z):
        self._coord = TCoordinates(x, y, z)

    def output(self, string):
        print('========', string, '========')
        print(f'\tx: {self._coord.x}\n\ty: {self._coord.y}\n\tz: {self._coord.z}\n')

    @property
    def coord(self):
        return self._coord

    def count_distace(self, t):
        pass


class TPoint_xy(TPoint):
    def __init__(self, x0, ax, y0, ay, t):
        x = x0 + ax * sin(t)
        y = y0 + ay * cos(t)
        z = 0
        super().__init__(x, y, z)

    def count_distace(self, t):
        return sqrt(pow(t.x - self._coord.x, 2) + pow(t.y - self._coord.y, 2))


class TPoint_xyz(TPoint):
    def __init__(self, x0, ax, y0, ay, z0, az, t):
        x = x0 + ax * sin(t)
        y = y0 + ay * cos(t)
        z = z0 + az * t * t
        super().__init__(x, y, z)

    def count_distace (self, t):
        return sqrt(pow(t.x - self._coord.x, 2) + pow(t.y - self._coord.y, 2) + pow(t.z - self._coord.z, 2))

File: Lab5/Cpp/TPoint.cpp
#include "TPoint.h"
#include <time.h>
#include <stdlib.h>
#include <math.h>
#include <iostream>
#include <string>
using namespace std;

double random_val()
{
	return (-10 + rand() % 20 + rand() / 10000.0);
}

void TPoint::output(string str)
{
	cout << "========" << str << "========" << endl;
	cout << "\tx: " << coord.get_x() << endl
		 << "\ty: " << coord.get_y() << endl
		 << "\tz: " << coord.get_z() << endl;
}

TPoint_xy::TPoint_xy(double t)
{
	double x0 = random_val();
	double ax = random_val();
	double y0 = random_val();
	double ay = random_val();
	coord.set_x(x0 + ax * sin(t));
	coord.set_y(y0 + ay * cos(t));
	coord.set_z(0);
}

TPoint_xy::TPoint_xy(double x0, double ax, double y0, double ay, double t)
{
	coord.set_x(x0 + ax * sin(t));
	coord.set_y(y0 + ay * cos(t));
	coord.set_z(0);
}

double TPoint_xy::count_distace(TCoordinates t)
{
	return sqrt(pow(t.get_x() - coord.get_x(), 2) + pow(t.get_y() - coord.get_y(), 2));
}

TPoint_xyz::TPoint_xyz(double t)
{
	double  x0 = random_val(),
			ax = random_val(),
			y0 = random_val(),
			ay = random_val(),
			z0 = random_val(),
			az = random_val();
	coord.set_x(x0 + ax * sin(t));
	coord.set_y(y0 + ay * cos(t));
	coord.set_z(z0 + az *  t * t);
}

TPoint_xyz::TPoint_xyz(double x0, double ax, double y0, double ay, double z0, double az, double t)
{
	coord.set_x(x0 + ax * sin(t));
	coord.set_y(y0 + ay * cos(t));
	coord.set_z(z0 + az *  t * t);
}

double TPoint_xyz::count_distace(TCoordinates t)
{
	return sqrt(pow(t.get_x() - coord.get_x(), 2) + pow(t.get_y() - coord.get_y(), 2) + pow(t.get_z() - coord.get_z(), 2));
}
